File: study/ts/script/scene/view/chat/ChatSystemMsgItemCell.ts
import ViewBase from "../../ViewBase";
import { Path } from "../../../utils/Path";
import { ChatConst } from "../../../const/ChatConst";
import UIHelper from "../../../utils/UIHelper";
import { Colors, G_ServerTime } from "../../../init";
import { table } from "../../../utils/table";
import { RichTextExtend } from "../../../extends/RichTextExtend";
import ListViewCellBase from "../../../ui/ListViewCellBase";
import RollNoticeHelper from "../rollnotice/RollNoticeHelper";

const {ccclass, property} = cc._decorator;

@ccclass
export default class ChatSystemMsgItemCell extends ListViewCellBase {

    @property({
        type: cc.Node,
        visible: true
    })
    _resourceNode: cc.Node = null;

    @property({
        type: cc.Sprite,
        visible: true
    })
    _imageChannel: cc.Sprite = null;

    @property({
        type: cc.Node,
        visible: true
    })
    _nodeMsg: cc.Node = null;

    @property(cc.Prefab)
    ChatTimeTipLayer:cc.Prefab = null;
    
    _chatMsg: any;
    _listWidth: any;
    _needShowTime: any;
    _extraHeight: number;
    _viewTimeNode: cc.Node;

    timeHeight:number = 0;
    richNode:cc.RichText;

    static YGAP = 16;
    static RICH_TEXT_MAX_WIDTH = 480;


    ctor(param) {
        var chatMsgData = param[0],listWidth = param[1];
        this._chatMsg = chatMsgData;
        this._listWidth = listWidth;
        this._needShowTime = this._chatMsg.getNeedShowTimeLabel();
        this._extraHeight = 0;
        this._updateUI();
    }
    _initUI() {
    }
    onCreate() {
        this._initUI();
        
    }
    onInit(){
        
    }
    onEnter(){

    }
    onExit(){

    }
    getTotalHeight() {
        var viewSize = this._resourceNode.getContentSize();
        var viewTimeHeight = this._viewTimeNode == null && 0 || this._viewTimeNode.getContentSize().height;
        return viewSize.height + viewTimeHeight + this._extraHeight;
    }
    _updateUI() {
        UIHelper.loadTexture(this._imageChannel, Path.getChatFaceMiniRes(ChatConst.CHANNEL_PNGS[this._chatMsg.getChannel()-1]));
        var msg = this._chatMsg.getSysMsg();
        var [_tmp,richElementList] = RollNoticeHelper.makeRichMsgFromServerRollMsg(msg, {
                textColor: '#'+Colors.BRIGHT_BG_TWO.toHEX('#rrggbb'),
                fontSize: 20
            });
        var element = this._createChannelRichElement(this._chatMsg.getChannel());
        var newElementList = [];
        table.insert(newElementList, element);
        for(let k in (richElementList as any)){
            newElementList.push(richElementList[k]);
        }
        var richStr = JSON.stringify(newElementList);
        this._showTxt(richStr);

        var timeNode = this._resourceNode.getChildByName("_viewTimeNode");
        if(timeNode)
        this._resourceNode.removeChild(timeNode,true);
        this._viewTimeNode = null;

        this.timeHeight = 0;
        var viewSize = this._resourceNode.getContentSize();
        if (this._needShowTime) {
            this._viewTimeNode = this._createTimeTipNode(G_ServerTime.getTimeString(this._chatMsg.getTime()));
            this._viewTimeNode.setPosition(this.node.width * 0.5, this.node.height);
            this._resourceNode.addChild(this._viewTimeNode);
            this.node.height = this.node.height + this._viewTimeNode.height;
            this._viewTimeNode.name = "_viewTimeNode";
            this.timeHeight = this._viewTimeNode.height;
        }
    }
    _createTimeTipNode(text):cc.Node {
        var viewTimeNode = cc.instantiate(this.ChatTimeTipLayer);
        var textTime = UIHelper.seekNodeByName(viewTimeNode, 'Text_Time');
        textTime.getComponent(cc.Label).string = (text);
        viewTimeNode.setAnchorPoint(0.5, 0);
        return viewTimeNode;
    }
    _showTxt(richStr) {
        var richTextMaxWidth = ChatSystemMsgItemCell.RICH_TEXT_MAX_WIDTH;
        var label = RichTextExtend.createWithContent(richStr);
        //label.setWrapMode(1);
        label.node.setAnchorPoint(cc.v2(0, 1));
        //label.setCascadeOpacityEnabled(true);
        //label.setVerticalSpace(3);
        //label.ignoreContentAdaptWithSize(false);
        //label.node.setContentSize(cc.size(richTextMaxWidth, 0));
        label.lineHeight = 26;
        label.maxWidth = richTextMaxWidth;
        //label.formatText();
        label.node.setPosition(0, 0);
        this._nodeMsg.removeAllChildren();
        this._nodeMsg.addChild(label.node);
        //logWarn('richtextHeight  --' + richtextHeight);
        var resourceSize = this._resourceNode.getContentSize();
        var totalHeight = label.node.getContentSize().height;
        this._extraHeight = Math.max(0, totalHeight - this._nodeMsg.y);
        this._nodeMsg.y = totalHeight;
       // this._resourceNode.y = (this._extraHeight);
        this.node.height = totalHeight;
        this.richNode = label;

    }
    _createChannelRichElement(channel) {
        var element = {} as any;
        element.type = 'image';
        element.filePath = Path.getChatFaceMiniRes(ChatConst.CHANNEL_PNGS[channel-1]);
        element.width = ChatConst.IMAGE_CHANNEL_WIDTH * ChatConst.IMAGE_CHANNEL_SCALE;
        element.height = ChatConst.IMAGE_CHANNEL_HEIGHT * ChatConst.IMAGE_CHANNEL_SCALE;
        return element;
    }
    getChatMsg() {
        return this._chatMsg;
    }
    updateItemSize(){
        if(!this.richNode){
            return;
        }
        var virtualContentSize = this.richNode.node.getContentSize();
        var richTextWidth = virtualContentSize.width;
        var richtextHeight = virtualContentSize.height;
        this._nodeMsg.y = richtextHeight;
        this.node.height = richtextHeight + this.timeHeight + 10;

        this.richNode = null;
    }
}

File: study/lua/src/app/scene/view/chat/ChatSystemMsgItemCell.lua
local CSHelper = require("yoka.utils.CSHelper")
local TypeConvertHelper = require("app.utils.TypeConvertHelper")
local RichTextHelper = require("app.utils.RichTextHelper")
local ViewBase = require("app.ui.ViewBase")
--单个聊天信息显示
local ChatSystemMsgItemCell = class("ChatSystemMsgItemCell", ViewBase)
local ChatConst = require("app.const.ChatConst")

ChatSystemMsgItemCell.YGAP = 16 -- 富文本上下间距
ChatSystemMsgItemCell.RICH_TEXT_MAX_WIDTH = 475

function ChatSystemMsgItemCell:ctor(chatMsgData, listWidth)
    self._chatMsg = chatMsgData --通知消息
    self._listWidth = listWidth --列表的宽度
    self._needShowTime = self._chatMsg:getNeedShowTimeLabel() --是否需要显示时间。
    self._resourceNode = nil
    self._viewTimeNode = nil
    self._imageChannel = nil
     --渠道图标
    self._extraHeight = 0 --聊天消息多出的显示高度
    self._nodeMsg = nil
    local resource = {
        file = Path.getCSB("ChatSystemMsgItemCell", "chat")
    }

    ChatSystemMsgItemCell.super.ctor(self, resource)
end

function ChatSystemMsgItemCell:_initUI()
end

function ChatSystemMsgItemCell:onCreate(...)
    self:_initUI()
    self:_updateUI()
end

function ChatSystemMsgItemCell:getTotalHeight()
    local viewSize = self._resourceNode:getContentSize()
    local viewTimeHeight = self._viewTimeNode == nil and 0 or self._viewTimeNode:getContentSize().height
    return viewSize.height + viewTimeHeight + self._extraHeight
end

function ChatSystemMsgItemCell:_updateUI()
    local ChatConst = require("app.const.ChatConst")
    self._imageChannel:loadTexture(Path.getTextSignet(ChatConst.CHANNEL_PNGS[self._chatMsg:getChannel()]))

    local RollNoticeHelper = require("app.scene.view.rollnotice.RollNoticeHelper")
    local msg = self._chatMsg:getSysMsg()
    local _,richElementList =
        RollNoticeHelper.makeRichMsgFromServerRollMsg(
        msg,
        {
            textColor = Colors.BRIGHT_BG_TWO,
            fontSize = 20
        }
    )
    local element = self:_createChannelRichElement(self._chatMsg:getChannel())
    table.insert(richElementList,1,element)
    local richStr = json.encode(richElementList)
    self:_showTxt(richStr)

    local viewSize = self._resourceNode:getContentSize()
    if self._needShowTime then
        self._viewTimeNode = self:_createTimeTipNode(G_ServerTime:getTimeString(self._chatMsg:getTime()))
        self._viewTimeNode:setPosition(viewSize.width * 0.5, viewSize.height)
        self._resourceNode:addChild(self._viewTimeNode)
    end
end

function ChatSystemMsgItemCell:_createTimeTipNode(text)
    local viewTimeNode = CSHelper.loadResourceNode(Path.getCSB("ChatTimeTipLayer", "chat"))
    local textTime = ccui.Helper:seekNodeByName(viewTimeNode, "Text_Time")
    textTime:setString(text)
    viewTimeNode:setAnchorPoint(0.5, 0)
    return viewTimeNode
end

---将文本转换为可以使用的富文本格式。
function ChatSystemMsgItemCell:_showTxt(richStr)
    local richTextMaxWidth = ChatSystemMsgItemCell.RICH_TEXT_MAX_WIDTH
     --self._listWidth - self._nodeMsg:getPositionX()-5
    local label = ccui.RichText:createWithContent(richStr)
    label:setWrapMode(1)
    label:setAnchorPoint(cc.p(0, 1))
    label:setCascadeOpacityEnabled(true)
    label:setVerticalSpace(3)
    --=======================================================
    --计算富文本的Size
    label:ignoreContentAdaptWithSize(false)
    label:setContentSize(cc.size(richTextMaxWidth, 0))
     --高度0则高度自适应
    label:formatText()
    local virtualContentSize = label:getVirtualRendererSize()
    local richTextWidth = virtualContentSize.width
    local richtextHeight = virtualContentSize.height
    --=======================================================
    label:setPosition(0, 0)

    self._nodeMsg:removeAllChildren()
    self._nodeMsg:addChild(label)

    logWarn("richtextHeight  --" .. richtextHeight)
    local resourceSize = self._resourceNode:getContentSize()
    self._extraHeight = math.max(0, richtextHeight - self._nodeMsg:getPositionY())
    self._resourceNode:setPositionY(self._extraHeight)
end

function ChatSystemMsgItemCell:_createChannelRichElement(channel)
    local element = {}
    element.type = "image"
    element.filePath = Path.getTextSignet(ChatConst.CHANNEL_PNGS[channel])
    element.width = ChatConst.IMAGE_CHANNEL_WIDTH * ChatConst.IMAGE_CHANNEL_SCALE
    element.height = ChatConst.IMAGE_CHANNEL_HEIGHT * ChatConst.IMAGE_CHANNEL_SCALE
    return element
end

function ChatSystemMsgItemCell:getChatMsg()
    return self._chatMsg
end

return ChatSystemMsgItemCell
